### ISSUE TITLE
Backfill blocks and state metadata

### ISSUE THREAD

**Issue by erikgrinaker** [tendermint/tendermint#4629]
Once we merge [state sync](#828) and [block pruning](#3652), it would be nice to be able to backfill historical blocks and associated state metadata (e.g. validator sets and commit info). See [RFC-001](https://github.com/tendermint/spec/pull/84) for discussion.

For state sync, this would allow apps to enforce a minimum block history invariant via e.g. a `backfill_height` metadata field in state sync snapshots and a `retain_height` ABCI commit response, and for block pruning this would allow a node with a truncated history to be converted into an archival node.

The existing blockchain reactor should probably be extended to handle this.

____

#### For Admin Use

- [x] Not duplicate issue
- [x] Appropriate labels applied
- [x] Appropriate contributors tagged
- [x] Contributor assigned/self-assigned

**Comment by cmwaters** [comment]
@erikgrinaker could I work on this? 

I see that there is already `retain_height` implemented. Is this something we want the application to be in control of though? I thought this would be left to the discretion of the individual nodes and that the application might instead enforce a minimum retain height.

In regards to the `backfill_height`, we would expect it to be set in the `StateSyncConfig` right? and the fast sync reactor would go backwards first and then forwards?

**Comment by erikgrinaker** [comment]
Sure, go for it.

We had a bunch of discussions about `retain_height` during the RFC review process (I think they're in the PR), and we wanted the app to control this since it's in a position to take all relevant parameters into account. It may be summarized in the RFC itself, can't remember. If the app wants to expose this to node operators, it can do so via its own config.

`backfill_height` should be a parameter of the snapshot metadata, not a configuration option (see e.g. https://github.com/tendermint/tendermint/blob/095e9cd7ef0e59c8b716a3bde32ca5b7a00f8e30/statesync/snapshots.go#L19-L27 and https://github.com/tendermint/tendermint/blob/095e9cd7ef0e59c8b716a3bde32ca5b7a00f8e30/proto/tendermint/statesync/types.proto#L17-L23). When the app generates a snapshot, it should specify the backfill height in it.

Actually implementing backfill may be a bit of a hassle. We probably want to reuse the fast sync reactor for this, but at the same time we don't want it to break if people disable the fast sync reactor. Also, this may mean that we'll have to implement support for this in all three versions of the fast sync reactor.

**Comment by cmwaters** [comment]
> Also, this may mean that we'll have to implement support for this in all three versions of the fast sync reactor.

I was thinking just for v0 and v2 since v1 will be deprecated in the next release right?

I just read over the RFC. Although I don't think we should reward poorly coded applications it would be a shame / somewhat funny if an application accidentally ordered all it's nodes to severe off it's history. 

> `backfill_height` should be a parameter of the snapshot metadata, not a configuration option - see e.g.:

And `backfill_height` would be set as the `retain_height` from the commit response at that height?

> We probably want to reuse the fast sync reactor for this, but at the same time we don't want it to break if people disable the fast sync reactor. 

I guess if people had it disabled we would just override it for this specific function and have it added somewhere here:

```golang	
           go func() {
		state, commit, err := ssR.Sync(stateProvider, config.DiscoveryTime)
		if err != nil {
			ssR.Logger.Error("State sync failed", "err", err)
			return
		}
		err = stateStore.Bootstrap(state)
		if err != nil {
			ssR.Logger.Error("Failed to bootstrap node with new state", "err", err)
			return
		}
		err = blockStore.SaveSeenCommit(state.LastBlockHeight, commit)
		if err != nil {
			ssR.Logger.Error("Failed to store last seen commit", "err", err)
			return
		}

                // add retrieval of past blocks here

		if fastSync {
			// FIXME Very ugly to have these metrics bleed through here.
			conR.Metrics.StateSyncing.Set(0)
			conR.Metrics.FastSyncing.Set(1)
			err = bcR.SwitchToFastSync(state)
			if err != nil {
				ssR.Logger.Error("Failed to switch to fast sync", "err", err)
				return
			}
		} else {
			conR.SwitchToConsensus(state, true)
		}
	}()
	return nil
```

Wouldn't be the cleanest of implementations though

**Comment by erikgrinaker** [comment]
>>Also, this may mean that we'll have to implement support for this in all three versions of the fast sync reactor.
>I was thinking just for v0 and v2 since v1 will be deprecated in the next release right?

I think we should either support it or remove it entirely, but not my decision.

>>backfill_height should be a parameter of the snapshot metadata, not a configuration option - see e.g.:
>And backfill_height would be set as the retain_height from the commit response at that height?

It's entirely up to the application how they want to populate `backfill_height`, since snapshots are generated by the application - Tendermint has no involvement in it. They may or may not be the same, depending on what the application needs. E.g. although retain height is set to 0 (keep everything) we may not want want to backfill the entire history.

This is something that should probably be configurable as well. I.e. even though the application only requests backfill of 1000 blocks, the operator may want to set up an archive node which backfills everything.

>>We probably want to reuse the fast sync reactor for this, but at the same time we don't want it to break if people disable the fast sync reactor.
>I guess if people had it disabled we would just override it for this specific function and have it added somewhere here:

Yeah, I dunno, this whole thing is just getting messier and messier - we should really get around to #4644. Also, the UX is pretty crap - like, the user explicitly disables fast sync, but we still run fast sync. It would be better to require fast sync to be enabled for using state sync, at least it's clearer to the user what is going on.

Something else we should consider: we may want block backfill to be asynchronous. If someone requests backfilling the entire block history, this is going to take hours or days, but we don't actually need to wait for these blocks - after state sync is done we can just start the node, and backfill while the node is running.

**Comment by cmwaters** [comment]
> Something else we should consider: we may want block backfill to be asynchronous.

But it needs to be synchronous in the sense that we should have the specified history before running fast sync or at least before running consensus (else we won't solve the problem of evidence). I agree for backfilling the entire history this needs to be asynchronous but I think we should almost make a distinction between the two forms of backfilling and perhaps address the "turning a statesynced node into an archival node" later. 

> It would be better to require fast sync to be enabled for using state sync, at least it's clearer to the user what is going on.

True, but I imagine this could also be confusing to users who on the cover see fast sync and state sync as two different ways of syncing. 

Another idea is that the backfilling could be executed in the state sync reactor alone but this would mean introducing blockstore and potentially block executor to the state sync which is currently quite a separate module so I'm not convinced of that. 

I would suggest that we still use the blockstore and define a method for backfilling blocks. We can draw the line here saying that state syncing requires used of the blockchain reactor but isn't fast syncing per se

**Comment by erikgrinaker** [comment]
>>Something else we should consider: we may want block backfill to be asynchronous.
>
> But it needs to be synchronous in the sense that we should have the specified history before running fast sync or at least before running consensus (else we won't solve the problem of evidence).

Yup. This should probably be what the application specifies in the snapshot, and if the node operator specifies additional backfilling that can be asynchronous.

>I agree for backfilling the entire history this needs to be asynchronous but I think we should almost make a distinction between the two forms of backfilling and perhaps address the "turning a statesynced node into an archival node" later.

I'd like to at least do the design work for this now, so that we know how we're going to have to implement it and won't have to redo a bunch of stuff later.

>Another idea is that the backfilling could be executed in the state sync reactor alone but this would mean introducing blockstore and potentially block executor to the state sync which is currently quite a separate module so I'm not convinced of that.

I don't see a problem with giving the state sync reactor access to the blockstore. If it needs it (to backfill blocks) then why shouldn't it have access to it?

The main issue is really that we'd be duplicating basically exactly what the fast sync reactor does (i.e. scheduling and fetching blocks from peers).

>I would suggest that we still use the blockstore and define a method for backfilling blocks. We can draw the line here saying that state syncing requires used of the blockchain reactor but isn't fast syncing per se.

This makes sense.

**Comment by cmwaters** [comment]
> if the node operator specifies additional backfilling that can be asynchronous.

How would you imagine this being done (the node specifying that they want to backfill more than what the app says)?

At the moment we run this upon receiving the `retain_height` from the application:

https://github.com/tendermint/tendermint/blob/095e9cd7ef0e59c8b716a3bde32ca5b7a00f8e30/consensus/state.go#L1613-L1618

Perhaps then, if the retain height was less than base, we could go back and fetch the missing blocks. This was if the application had some local config that it accounted for then the node operator could just change this if he/she wanted to backfill more blocks

**Comment by erikgrinaker** [comment]
The application may or may not expose configuration options to control it. Even though the application defaults to keeping e.g. the last 1000 blocks, an operator wanting to run an archive node would enable full block retention in the application config.

`retain_height` and `backfill_height` are logically orthogonal concepts, although we'll need to consider their interactions. For example, if the application returns a `retain_height` that is above the current `backfill_height`, it should cause a running backfill to increase its backfill height (since the blocks that are being backfilled have now been pruned, whether or not they actually exist yet). It is up to application developers and node operators to configure these correctly.

**Comment by cmwaters** [comment]
> retain_height and backfill_height are logically orthogonal concepts

Can you explain their discrepancies? They way I see it currently it could be perfectly possible to have an implementation with just a single variable dictating where the base of the local blockchain should be (especially if we're passing responsibility of the decision to the application).  Perhaps the only distinction I can think of is whether the operation should be synchronous or not.

**Comment by erikgrinaker** [comment]
> Can you explain their discrepancies? 

In the typical case, `retain_height` would be 0 (retain everything), but a node operator doing state sync would presumably not want to backfill the entire history, but rather the minimum necessary for e.g. evidence to work (i.e. 3 weeks worth of blocks). If they're backfilling everything, then what's the point of state sync?

**Comment by cmwaters** [comment]
Okay, so we want to encompass two attributes:
 a) we want to set a minimum amount of blocks that the application and tendermint require to operate and 
 b) we want to allow nodes to locally choose how many blocks they keep so long as it's greater than (a). It's important for auditing / start up that some nodes have the full history.

Now we could give complete responsibility to the app (which we have already done) through `retain_height` which could account for both of these concerns and would thus assume that their is some local config that node operators can choose if they want to backfill more than necessary.

Or we could have two variables, the minimum amount of blocks or `retain_height` set by the app and another variable controlled by the node operator directly. This would probably need to be set by the node, most likely in the config and must be greater than the retain height or if it wasn't then tendermint would use the retain height to know at which height to set the base of the blockchain.

**Comment by erikgrinaker** [comment]
>Now we could give complete responsibility to the app (which we have already done) through retain_height which could account for both of these concerns and would thus assume that their is some local config that node operators can choose if they want to backfill more than necessary.

Except we can't really do this, because `retain_height` is returned after we commit a block, but when backfilling we need to fetch at blocks before executing anything (because we need these e.g. if there is evidence in the block), and thus we don't know what `retain_height` is yet. The backfill height must be part of the snapshot metadata, since the application knows the minimum number of blocks to keep (because it knows e.g. the unbonding time).

We may want to have an _additional_ config option to allow node operators to _extend_ the backfill height (e.g. if the app requires 3 weeks, but the operator wants to backfill everything). This should be in addition to what the snapshot metadata specifies.

> Or we could have two variables, the minimum amount of blocks or retain_height set by the app and another variable controlled by the node operator directly.

For `retain_height`, we already discussed this during the RFC process. If there should be a configuration option to extend `retain_height`, it should be part of the application config, so that the application can take it into account and compare it with other required parameters (such as IAVL persist intervals, snapshot intervals, and unbonding time).

**Comment by alexanderbez** [comment]
Note, the minimum `retain_height` is already an app configuration of the SDK. See [here](https://github.com/cosmos/cosmos-sdk/blob/6344d626db1fbdba5e0f67425703c1584021bf5b/baseapp/abci.go#L623) for how we determine `retain_height` in the context of the SDK.

**Comment by cmwaters** [comment]
So `backfill_height` is really just a temporary value that is used in start-up but doesn't persist once the node has backfilled to that height and can start consensus.

**Comment by erikgrinaker** [comment]
Just remembered that we'll need to backfill state metadata as well, e.g. what validator sets and consensus params were at specific heights. With the current fast sync we actually execute blocks and get validator set updates from the ABCI application, but with backfill the blocks won't be executed and so we'll need to fetch this via the blockchain reactor with new message types.

>So backfill_height is really just a temporary value that is used in start-up but doesn't persist once the node has backfilled to that height and can start consensus (and from then on out it uses retain_height).

Yeah, that's how I think about it at least. Note that backfill may be aborted (e.g. if the process restarts), in which case it should pick back up again when it restarts, so it does need _some_ form of persistence to track progress. But once backfill is done, backfill_height can be discarded I think.

**Comment by cmwaters** [comment]
> Just remembered that we'll need to backfill state metadata as well, e.g. what validator sets and consensus params were at specific heights. With the current fast sync we actually execute blocks and get validator set updates from the ABCI application, but with backfill the blocks won't be executed and so we'll need to fetch this via the blockchain reactor with new message types.

True. This shouldn't be too hard to set up though since we have access to the state store through the block executor and we can verify each of these with the header we validate. 

Do you think any more consideration for design is necessary? I think I have fairly good grasp of what needs to be implemented so I can start with it tomorrow. 

Btw adding validator set messages to the blockchain reactor means we basically have all the ingredients to support p2p light clients

**Comment by erikgrinaker** [comment]
This will need an RFC in the spec repo, since it requires protocol changes - Informal probably wants a say as well.

**Comment by erikgrinaker** [comment]
Eh, on that note: the state sync reactor already has access to this state data via the light client, of course, so we can just use that to backfill state data.

**Comment by cmwaters** [comment]
but just the validator sets, not the consensus params nor abci responses

**Comment by erikgrinaker** [comment]
Params are exposed via RPC anyway, so should be easy to thread them through the light client and state provider. As for ABCI responses, there's been some talk of getting rid of them entirely, see #3533. May not be critical to sync, but would be nice to get a final decision on this, and if it _is_ something we want to keep around then expose them via RPC as well.

**Comment by cmwaters** [comment]
> ... so we can just use that to backfill state data.

Would we want to run two different operations then; one to fill state data and one for block data?

**Comment by erikgrinaker** [comment]
Probably, but I think we should make sure state data is available before block data - especially for async backfill. Otherwise this may confuse RPC clients that assume a node with a given height also has state for that height (e.g. state sync itself assumes this from its RPC servers).

**Comment by cmwaters** [comment]
The decision was made in RFC005 to not implement backfill of all state and block data rather to just have backfill of necessary data for evidence handling which would immediately follow a state sync. 

Operators wishing to create an archive node can always fast sync a new node from genesis
